keps/sig-windows/20190714-windows-csi-support.md

```diff
@@ -0,0 +1,401 @@
+
+A CSI Node Plugin will interact with csi-proxy.exe using a named pipe: `\\.\pipe\csi-proxy` (exposed by csi-proxy.exe). The pipe will need to be mounted into the Node Plugin container from the host using the pod's volume mount specifications. Note that domain sockets cannot be used in this scenario since Windows blocks a containerized process from interacting with a host process that is listening on a domain socket. 
+
+A GRPC based interface will be used by CSI Node Plugins to invoke privileged operations through csi-proxy.exe. The GRPC API will be versioned and any release of csi-proxy.exe binary will maintain backward compatibility across all prior versions of the API. This avoids having to run multiple versions of the csi-proxy.exe binary on the same Windows host if multiple CSI Node Plugins have been configured and the version of the csi-proxy API required by the plugins are wide apart.
+
+The following are the main RPC calls that will comprise a v1 version of the csi-proxy API:
+
+```
+service CSIProxyService {
+
+    // Version returns the earliest and latest version of the CSI Proxy API supported by the service.
+    rpc Version(VersionRequest) returns (VersionResponse) {}
+
+    // Mkdir creates a directory at the requested absolute path in the host. Relative path is not supported.
+    rpc Mkdir(MkdirRequest) returns (MkdirResponse) {}
+
+    // Rmdir removes the directory at the requested absolute path in the host. Relative path is not supported.
+    // This may be used for unlinking a symlink created through LinkVolume
+    rpc Rmdir(RmdirRequest) returns (RmdirResponse) {}
+
+    // StageDisk partitions a disk device and formats the partition with NTFS (if the disk has not been partitioned or formatted)
+    // The resulting volume is mounted at the requested global staging path.
+    rpc StageDisk(StageDiskRequest) returns (StageDiskResponse) {}
+
+    // StageSMBShare mounts a remote share over SMB on the host at the requested global staging path.
+    rpc StageSMBShare(StageSMBShareRequest) returns (StageSMBShareResponse) {}
+
+    // LinkVolume invokes mklink on the global staging path of a volume linking it to a path within a container
+    rpc LinkVolume(LinkVolumeRequest) returns (LinkVolumeResponse) {}
+
+    // GetDiskNumberWithLocation finds a disk device with requested Location <Bus, Target, LUN ID>
+    rpc GetDiskNumberWithLocation(GetDiskNumberWithLocationRequest) returns (GetDiskNumberWithLocationResponse) {}
+
+    // GetDiskNumberWithSCSIID finds a disk device whose SCSI Page 83 ID contains the passed ID string
+    rpc GetDiskNumberWithSCSIID(GetDiskNumberWithSCSIIDRequest) returns (GetDiskNumberWithSCSIIDResponse) {}
+}
+```
+
+In accordance with standard Kubernetes conventions, the above API will be introduces as v1alpha1 and graduate to v1beta1 and v1 as the feature graduates. New RPCs and enhancements to parameters will be introduced through v1alpha1 and graduate to v1beta1 and v1 versions as they mature.
```

Comment:
How would the the development and release workflow look like when CSI is updated and a new RPC call is needed in the proxy service?

Reply:
A good example of this is the new Resize API in the node plugin interface which requires a ResizeParition operation on the node plugin interface as well. In such cases, we will have to fast-track the API to v1 in the proxy. Given the proxy will be an out-of-tree component, such fast-track exceptions should be possible. I will detail this out here along with the need to anticipate potential changes in the proxy interface when CSI changes are considered.